pysph/sph/wc/crksph.py

```diff
@@ -373,3 +384,468 @@ def loop(self, d_idx, s_idx, d_ai, d_gradai, d_cwij, d_bi, d_gradbi, s_ai,
+
+    .. math::
+        \mu_i = \min \left(0, \frac{\hat{v}_{ij}
+        \eta_{i}^{\alpha}}{\eta_{i}^{\alpha}\eta_{i}^{\alpha}
+        + \epsilon^{2}}\right)
+
+    .. math::
+        \hat{v}_{ij}^{\alpha} = v_{i}^{\alpha} - v_{j}^{\alpha}
+        - \frac{\phi_{ij}}{2}\left(\partial_{\beta} v_i^{\alpha}
+        + \partial_{\beta}v_j^{\alpha}\right) x_{ij}^{\beta}
+
+    .. math::
+        \phi_{ij} = \max \left[0, \min \left[1, \frac{4r_{ij}}{(1
+        + r_{ij})^2}\right]\right] \times
+        \begin{cases}
+        \exp{\left[-\left((\eta_{ij}
+        - \eta_{crit})/\eta_{fold}\right)^2\right]}, &\eta_{ij} < \eta_{crit}
+        \\
+        1, &  \eta_{ij} >= \eta_{crit}
+        \end{cases}
+
+    .. math::
+        \eta_{ij} = \min(\eta_i, \eta_j)
+
+    .. math::
+        \eta_{i/j} = (x_{ij}^{\alpha} x_{ij}^{\alpha})^{1/2} / h_{i/j}
+
+    .. math::
+        r_{ij} = \frac{\partial_{\beta} v_i^{\alpha} x_{ij}^{\alpha}
+        x_{ij}^{\beta}}{\partial_{\beta} v_j^{\alpha}x_{ij}^{\alpha}
+        x_{ij}^{\beta}}
+
+    .. math::
+        \partial_{\beta} v_i^{\alpha} = -\sum_j V_j v_{ij}^{\alpha}
+        \partial_{\beta} W_{ij}^R
+    """
+
+    def __init__(self, dest, sources, dim, gx=0.0, gy=0.0, gz=0.0, cl=2, cq=1,
+                 eta_crit=0.5, eta_fold=0.2, tol=0.5):
+        self.dim = dim
+        self.gx = gx
+        self.gy = gy
+        self.gz = gz
+        self.cl = cl
+        self.cq = cq
+        self.eta_crit = eta_crit
+        self.eta_fold = eta_fold
+        self.tol = tol
+        super(MomentumEquation, self).__init__(dest, sources)
+
+    def _get_helpers_(self):
+        return [dot]
+
+    def initialize(self, d_idx, d_au, d_av, d_aw):
+        d_au[d_idx] = self.gx
+        d_av[d_idx] = self.gy
+        d_aw[d_idx] = self.gz
+
+    def loop(self, d_idx, s_idx, d_m, s_m, d_rho, s_rho, d_p, s_p, d_cs, s_cs,
+             d_u, d_v, d_w, s_u, s_v, s_w, d_gradv, s_gradv, d_h, s_h, s_ai,
+             s_bi, s_gradai, s_gradbi, d_au, d_av, d_aw, d_V, s_V, XIJ, VIJ,
+             DWIJ, EPS, HIJ, WIJ):
+        Cl = self.cl
+        Cq = self.cq
+        eta_fold = self.eta_fold
+        eta_crit = self.eta_crit
+
+        mi = d_m[d_idx]
+        Vi = 1.0/d_V[d_idx]
+        Vj = 1.0/s_V[s_idx]
+
+        pi = d_p[d_idx]
+        pj = s_p[s_idx]
+
+        rhoi = d_rho[d_idx]
+        rhoj = s_rho[s_idx]
+
+        ci = d_cs[d_idx]
+        cj = s_cs[s_idx]
+
+        hi = d_h[d_idx]
+        hj = s_h[s_idx]
+
+        alp, bet, i, d = declare('int', 4)
+        uijhat, tmpdvxij = declare('matrix(3)', 2)
+
+        d = self.dim
+
+        tmpri = 0.0
+        tmprj = 0.0
+        for alp in range(d):
+            for bet in range(d):
+                tmpri += d_gradv[d*d*d_idx + d*alp + bet] * XIJ[alp] * XIJ[bet]
+                tmprj += s_gradv[d*d*s_idx + d*alp + bet] * XIJ[alp] * XIJ[bet]
+        rij = tmpri/tmprj
+
+        tmprij = min(1, 4*rij/(1 + rij)**2.0)
```

Comment:
Please use (1 + rij)*(1+rij) as when we transpile this, it will use "pow" which could be much less efficient.

Reply:
fixed it, thanks!